jvm/sargon-android/src/main/java/com/radixdlt/sargon/os/storage/EncryptionHelper.kt

```diff
@@ -0,0 +1,146 @@
+@file:OptIn(ExperimentalEncodingApi::class)
+
+package com.radixdlt.sargon.os.storage
+
+import com.radixdlt.sargon.extensions.then
+import java.nio.ByteBuffer
+import java.nio.charset.StandardCharsets
+import java.security.SecureRandom
+import java.security.spec.AlgorithmParameterSpec
+import javax.crypto.Cipher
+import javax.crypto.SecretKey
+import javax.crypto.spec.GCMParameterSpec
+import javax.crypto.spec.SecretKeySpec
+import kotlin.io.encoding.Base64
+import kotlin.io.encoding.ExperimentalEncodingApi
+
+internal object EncryptionHelper {
+    fun encrypt(
+        input: ByteArray,
+        secretKey: SecretKey
+    ): Result<ByteArray> {
+        return runCatching {
+            val cipher = Cipher.getInstance(AES_GCM_NOPADDING)
+            val ivBytes = ByteArray(GCM_IV_LENGTH)
+            SecureRandom().nextBytes(ivBytes)
+            val parameterSpec = GCMParameterSpec(AUTH_TAG_LENGTH, ivBytes)
+            cipher.init(Cipher.ENCRYPT_MODE, secretKey, parameterSpec)
+
+            val ciphertext: ByteArray = cipher.doFinal(input)
+
+            val byteBuffer = ByteBuffer.allocate(ivBytes.size + ciphertext.size)
+            byteBuffer.put(ivBytes)
+            byteBuffer.put(ciphertext)
+
+            byteBuffer.array()
+        }
+    }
+
+    fun decrypt(
+        input: ByteArray,
+        secretKey: SecretKey
+    ): Result<ByteArray> {
+        return runCatching {
+            val cipher = Cipher.getInstance(AES_GCM_NOPADDING)
+            val gcmIv: AlgorithmParameterSpec =
+                GCMParameterSpec(AUTH_TAG_LENGTH, input, 0, GCM_IV_LENGTH)
+            cipher.init(Cipher.DECRYPT_MODE, secretKey, gcmIv)
+            cipher.doFinal(input, GCM_IV_LENGTH, input.size - GCM_IV_LENGTH)
+        }
+    }
+
+    const val AES_ALGORITHM = "AES"
+    private const val AES_GCM_NOPADDING = "AES/GCM/NoPadding"
+    private const val GCM_IV_LENGTH = 12
+    private const val AUTH_TAG_LENGTH = 128 // bit
+}
+
+/**
+ * Encrypts this value with the provided [secretKey].
+ *
+ * The receiver must be either a [String] or a [ByteArray]. Other types are not supported as
+ * of this moment.
+ */
+@Suppress("UNCHECKED_CAST")
+internal fun <T : Any> T.encrypt(secretKey: SecretKey): Result<T> = runCatching {
+    when (this) {
+        is String -> Base64.encode(
+            source = EncryptionHelper.encrypt(
+                input = toByteArray(),
+                secretKey = secretKey
+            ).getOrThrow()
+        ) as T
+
+        is ByteArray -> EncryptionHelper.encrypt(input = this, secretKey = secretKey)
+            .getOrThrow() as T
+
+        else -> throw UnsupportedOperationException(
```

Comment:
minor but wouldn't be more a `IllegalArgumentException` here?

Reply:
Right will change that